validator/src/main/java/com/google/daq/mqtt/sequencer/sequences/ConfigSequences.java

```diff
@@ -116,10 +129,151 @@ public void extra_config() {
+   * then setting the config.pointset.sample_rate_min to match half the initial interval
+   * and measuring the final interval between several messages and ensuring it is less than the new
+   * interval (with a tolerance of 1.5s).
+   *
+   * <p>Pass if: final interval < new sample_rate_min + tolerance
+   * Fail if: final interval > new sample_rate_min + tolerance
+   * Skip if: initial interval < 5s (too fast for automated test)
+   */
+  @Test(timeout = 180000)
+  @Description("device publishes pointset events at a rate of no more than config sample_rate_sec")
+  public void pointset_sample_rate() {
+    Integer defaultSampleRate = 10;
+
+    // Clear received events because this could contain messages from a previous sample rate test
+    getReceivedEvents(PointsetEvent.class);
+
+    Instant endTime = Instant.now().plusSeconds(defaultSampleRate * 3);
+    // To pick the test sample rate, either measure the devices
+    // given sampling rate from its last 2 messages and half it or use
+    // a value if long
+    untilTrue("measure initial sample rate",
+        () -> (countReceivedEvents(PointsetEvent.class) > 1
+          || Instant.now().isAfter(endTime))
+    );
+
+    Integer testSampleRate;
+    if (countReceivedEvents(PointsetEvent.class) < 2) {
+      // 2 messages not seen, assume interval is longer than wait period, pick a small number
+      testSampleRate = defaultSampleRate;
+    } else {
+      List<PointsetEvent> receivedEvents = getReceivedEvents(PointsetEvent.class);
+      List<Long> telemetryDelta = intervalFromEvents(receivedEvents);
+      Integer nominalInterval = telemetryDelta.get(0).intValue();
+      info(String.format("initial sample rate is %d seconds", nominalInterval));
+
+      if (nominalInterval < 5) {
+        throw new SkipTest("measured sample rate is too low for automated test");
+      }
+
+      // Use an interval smaller than the devices last interval
+      testSampleRate = Math.floorDiv(nominalInterval, 2);
+    }
+
+    info(String.format("setting sample rate to %d seconds", testSampleRate));
+    SamplingRange testSampleRange = new SamplingRange(1, testSampleRate, 1.5);
+    checkThat(samplingMessagesCheckMessage(testSampleRange),
+        () -> testPointsetWithSamplingRange(testSampleRange, 5, 2)
+    );
+  }
+
+  /**
+   * Generates message for checking the time periods are within the sampling range.
+   *
+   * @param samplingRange sampling range to produce message for
+   * @return message
+   */
+  private String samplingMessagesCheckMessage(SamplingRange samplingRange) {
+    return String.format("time period between successive pointset events is %s",
+        samplingRange);
+  }
+
+  /**
+   * Tests both sample_rate_sec and sample_limit_sec by defining two non-intersecting narrow
+   * ranges of both parameters, and ensuring telemetry is within this range.
+   */
+  @Test(timeout = 240000)
```

Comment:
is it possible to make the 240000 a NAMED_CONSTANT_MS?  I'm not sure syntactically if it's allowed, but would help readability overall!

Reply:
Yes it works, added a class of constants e.g. `FOUR_MINUTES_MS`. Let me know that needs changing